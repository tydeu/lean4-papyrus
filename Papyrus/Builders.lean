import Papyrus.Context
import Papyrus.IR.ModuleRef
import Papyrus.IR.FunctionRef
import Papyrus.IR.GlobalVariableRef
import Papyrus.IR.InstructionRefs

namespace Papyrus

-- # Module Builder

abbrev ModuleM := ReaderT ModuleRef LlvmM

protected def ModuleM.runIn (mod : ModuleRef) (self : ModuleM α) : LlvmM α :=
  self mod

-- # Basic Block Builder

structure BasicBlockContext where
  modRef : ModuleRef
  funRef : FunctionRef
  bbRef : BasicBlockRef

abbrev BasicBlockM := ReaderT BasicBlockContext LlvmM

instance : MonadLift ModuleM BasicBlockM where
  monadLift m := fun ctx => m ctx.modRef

protected def BasicBlockM.runIn (ctx : BasicBlockContext) (self : BasicBlockM α) : LlvmM α :=
  self ctx

namespace Builder

def module (name : String) (builder : ModuleM PUnit) : LlvmM ModuleRef := do
  let modRef ← ModuleRef.new name
  builder.runIn modRef
  return modRef

-- ## Module Builder Actions

/-- Add a global variable to the module. -/
def globalVar (type : TypeRef)
(isConstant := false) (linkage := Linkage.external) (name : @& String := "")
(tlm := ThreadLocalMode.notLocal) (addrSpace := AddressSpace.default) (isExternallyInitialized := false)
: ModuleM GlobalVariableRef := do
  let gblRef ← GlobalVariableRef.new type isConstant linkage name tlm addrSpace isExternallyInitialized
  (← read).appendGlobalVariable gblRef
  return gblRef

/-- Add a global variable with an initializer to the module. -/
def globalVarInit (init : ConstantRef)
(isConstant := false) (linkage := Linkage.external) (name : @& String := "")
(tlm := ThreadLocalMode.notLocal) (addrSpace := AddressSpace.default)
: ModuleM GlobalVariableRef := do
  let gblRef ← GlobalVariableRef.ofConstant init isConstant linkage name tlm addrSpace
  (← read).appendGlobalVariable gblRef
  return gblRef

/-- Add a string constant to the module. -/
def string (str : String)
(addrSpace := AddressSpace.default) (withNull := true) (name := "")
: ModuleM GlobalVariableRef := do
  let gblRef ← GlobalVariableRef.ofString str addrSpace withNull name
  (← read).appendGlobalVariable gblRef
  return gblRef

/-- Add a string constant to the module and return a constant pointer to its head. -/
def stringPtr (str : String)
(addrSpace := AddressSpace.default) (withNull := true) (name := "")
: ModuleM ConstantRef := do
  let gblRef ← string str addrSpace withNull name
  let zeroRef ← ConstantIntRef.ofUInt32 0
  let ptrRef ← ConstantExprRef.getGetElementPtr gblRef #[zeroRef, zeroRef] true
  return ptrRef

/-- Add a arbitrary constant to the module. -/
def globalConst (init : ConstantRef)
(linkage := Linkage.linkOnceODR) (name := "")
(tlm := ThreadLocalMode.notLocal) (addrSpace := AddressSpace.default) : ModuleM GlobalVariableRef := do
  let gblRef ← GlobalVariableRef.ofConstant init true linkage name tlm addrSpace
  (← read).appendGlobalVariable gblRef
  return gblRef

/-- Add a function declaration to the module. -/
def declare (type : FunctionTypeRef) (name : String)
(linkage := Linkage.external) (addrSpace := AddressSpace.default) : ModuleM FunctionRef := do
  let funRef ← FunctionRef.create type name linkage addrSpace
  (← read).appendFunction funRef
  return funRef

/-- Define a new a function in the module. -/
def define (type : FunctionTypeRef) (builder : BasicBlockM PUnit) (name : String :=  "")
(linkage := Linkage.external) (addrSpace := AddressSpace.default) (entry : String := "") : ModuleM FunctionRef := do
  let funRef ← FunctionRef.create type name linkage addrSpace
  let bbRef ← BasicBlockRef.create entry
  funRef.appendBasicBlock bbRef
  let modRef ← read
  modRef.appendFunction funRef
  builder.runIn {modRef, funRef, bbRef}
  return funRef

-- ## Basic Block Builder Actions

def getArg (argNo : UInt32) : BasicBlockM ArgumentRef := do
  (← read).funRef.getArg argNo

def label (name : String) (builder : BasicBlockM PUnit) : BasicBlockM BasicBlockRef := do
  let ctx ← read
  let bb ← BasicBlockRef.create name
  ctx.funRef.appendBasicBlock bb
  builder.runIn {ctx with bbRef := bb}
  return bb

-- ### `ret`

def retVoid : BasicBlockM PUnit := do
  (← read).bbRef.appendInstruction <| ← ReturnInstRef.createVoid

def ret (val : ValueRef)  : BasicBlockM PUnit := do
  (← read).bbRef.appendInstruction <| ← ReturnInstRef.create val

-- ### `br`

def condBr (cond : ValueRef) (ifTrue ifFalse : BasicBlockRef)  : BasicBlockM PUnit := do
  (← read).bbRef.appendInstruction <| ← CondBrInstRef.create ifTrue ifFalse cond

def br (bb : BasicBlockRef)  : BasicBlockM PUnit := do
  (← read).bbRef.appendInstruction <| ← BrInstRef.create bb

-- ### `load`

def load (type : TypeRef) (ptr : ValueRef) (name := "") (isVolatile := false)
  (align : Align := 1) (order := AtomicOrdering.notAtomic) (ssid := SyncScopeID.system)
  : BasicBlockM InstructionRef := do
  let inst ← LoadInstRef.create type ptr name isVolatile align order ssid
  (← read).bbRef.appendInstruction inst
  return inst

-- ### `store`

def store (val : ValueRef) (ptr : ValueRef) (isVolatile := false)
  (align : Align := 1) (order := AtomicOrdering.notAtomic) (ssid := SyncScopeID.system)
  : BasicBlockM InstructionRef := do
  let inst ← StoreInstRef.create val ptr isVolatile align order ssid
  (← read).bbRef.appendInstruction inst
  return inst

-- ### `getelementptr`

def getElementPtr
(pointeeType : TypeRef) (ptr : ValueRef) (indices : Array ValueRef := #[])
(name : String := "") : BasicBlockM InstructionRef := do
  let inst ← GetElementPtrInstRef.create pointeeType ptr indices name
  (← read).bbRef.appendInstruction inst
  return inst

def getElementPtrInbounds
(pointeeType : TypeRef) (ptr : ValueRef) (indices : Array ValueRef := #[])
(name : String := "") : BasicBlockM InstructionRef := do
  let inst ← GetElementPtrInstRef.createInbounds pointeeType ptr indices name
  (← read).bbRef.appendInstruction inst
  return inst

-- ### `call`

def call (fn : FunctionRef) (args : Array ValueRef := #[]) (name : String := "") : BasicBlockM InstructionRef := do
  let inst ← fn.createCall args name
  (← read).bbRef.appendInstruction inst
  return inst

def callAs (type : FunctionTypeRef) (fn : ValueRef) (args : Array ValueRef := #[]) (name : String := "") : BasicBlockM InstructionRef := do
  let inst ← CallInstRef.create type fn args name
  (← read).bbRef.appendInstruction inst
  return inst

-- ## Binary Operators

def binop (op : InstructionKind) (s1 s2 : ValueRef) (name : String := "") (h : op.is_binary_op := by trivial) : BasicBlockM BinaryOperatorRef := do
  let inst ← BinaryOperatorRef.create op s1 s2 name h
  (← read).bbRef.appendInstruction inst
  return inst

def add (s1 s2 : ValueRef) (name : String := "") : BasicBlockM BinaryOperatorRef := do
  let inst ← BinaryOperatorRef.create InstructionKind.add s1 s2 name
  (← read).bbRef.appendInstruction inst
  return inst

def fadd (s1 s2 : ValueRef) (name : String := "") : BasicBlockM BinaryOperatorRef := do
  let inst ← BinaryOperatorRef.create InstructionKind.fadd s1 s2 name
  (← read).bbRef.appendInstruction inst
  return inst

def sub (s1 s2 : ValueRef) (name : String := "") : BasicBlockM BinaryOperatorRef := do
  let inst ← BinaryOperatorRef.create InstructionKind.sub s1 s2 name
  (← read).bbRef.appendInstruction inst
  return inst

def fsub (s1 s2 : ValueRef) (name : String := "") : BasicBlockM BinaryOperatorRef := do
  let inst ← BinaryOperatorRef.create InstructionKind.fsub s1 s2 name
  (← read).bbRef.appendInstruction inst
  return inst

def mul (s1 s2 : ValueRef) (name : String := "") : BasicBlockM BinaryOperatorRef := do
  let inst ← BinaryOperatorRef.create InstructionKind.mul s1 s2 name
  (← read).bbRef.appendInstruction inst
  return inst

def fmul (s1 s2 : ValueRef) (name : String := "") : BasicBlockM BinaryOperatorRef := do
  let inst ← BinaryOperatorRef.create InstructionKind.fmul s1 s2 name
  (← read).bbRef.appendInstruction inst
  return inst

def udiv (s1 s2 : ValueRef) (name : String := "") : BasicBlockM BinaryOperatorRef := do
  let inst ← BinaryOperatorRef.create InstructionKind.udiv s1 s2 name
  (← read).bbRef.appendInstruction inst
  return inst

def sdiv (s1 s2 : ValueRef) (name : String := "") : BasicBlockM BinaryOperatorRef := do
  let inst ← BinaryOperatorRef.create InstructionKind.sdiv s1 s2 name
  (← read).bbRef.appendInstruction inst
  return inst

def fdiv (s1 s2 : ValueRef) (name : String := "") : BasicBlockM BinaryOperatorRef := do
  let inst ← BinaryOperatorRef.create InstructionKind.fdiv s1 s2 name
  (← read).bbRef.appendInstruction inst
  return inst

def urem (s1 s2 : ValueRef) (name : String := "") : BasicBlockM BinaryOperatorRef := do
  let inst ← BinaryOperatorRef.create InstructionKind.urem s1 s2 name
  (← read).bbRef.appendInstruction inst
  return inst

def srem (s1 s2 : ValueRef) (name : String := "") : BasicBlockM BinaryOperatorRef := do
  let inst ← BinaryOperatorRef.create InstructionKind.srem s1 s2 name
  (← read).bbRef.appendInstruction inst
  return inst

def frem (s1 s2 : ValueRef) (name : String := "") : BasicBlockM BinaryOperatorRef := do
  let inst ← BinaryOperatorRef.create InstructionKind.frem s1 s2 name
  (← read).bbRef.appendInstruction inst
  return inst

def shl (s1 s2 : ValueRef) (name : String := "") : BasicBlockM BinaryOperatorRef := do
  let inst ← BinaryOperatorRef.create InstructionKind.shl s1 s2 name
  (← read).bbRef.appendInstruction inst
  return inst

def lshr (s1 s2 : ValueRef) (name : String := "") : BasicBlockM BinaryOperatorRef := do
  let inst ← BinaryOperatorRef.create InstructionKind.lshr s1 s2 name
  (← read).bbRef.appendInstruction inst
  return inst

def ashr (s1 s2 : ValueRef) (name : String := "") : BasicBlockM BinaryOperatorRef := do
  let inst ← BinaryOperatorRef.create InstructionKind.ashr s1 s2 name
  (← read).bbRef.appendInstruction inst
  return inst

def and (s1 s2 : ValueRef) (name : String := "") : BasicBlockM BinaryOperatorRef := do
  let inst ← BinaryOperatorRef.create InstructionKind.and s1 s2 name
  (← read).bbRef.appendInstruction inst
  return inst

def or (s1 s2 : ValueRef) (name : String := "") : BasicBlockM BinaryOperatorRef := do
  let inst ← BinaryOperatorRef.create InstructionKind.or s1 s2 name
  (← read).bbRef.appendInstruction inst
  return inst

def xor (s1 s2 : ValueRef) (name : String := "") : BasicBlockM BinaryOperatorRef := do
  let inst ← BinaryOperatorRef.create InstructionKind.xor s1 s2 name
  (← read).bbRef.appendInstruction inst
  return inst
