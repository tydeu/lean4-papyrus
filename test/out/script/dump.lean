import Papyrus

open Papyrus Script

-- # Module

llvm module hello do
  declare i8 @printf(i8*, ...)
  define i32 @main() do
    call @printf("Hello World"*)
    ret i32 0

#dump hello

-- # Types

#dump llvm type void
#dump llvm type i32
#dump llvm type i8 (i8*, ...)
#dump llvm type i8 addrspace(5)*
#dump llvm type {i32, float}
#dump llvm type [2 x i8]
#dump llvm type <4 x i64>
#dump llvm type <vscale x 4 x double>

-- # Constants

#dump llvm true
#dump llvm false
#dump llvm i8 255
#dump llvm i32 1
#dump llvm i64 -1
#dump llvm i128 1208925819614629174706188 -- 2^80 + 12
#dump llvm i128 -1208925819614629174706188


-- # Ops

llvm module ops do
  define i32 @opsEx(i32 %a, i32 %b) do
    %c = add i32 %a, %b
    %d = mul i32 %a, %b
    %e = sub i32 %a, %b
    %f = udiv i32 %a, %b 
    %g = sdiv i32 %a, %b 
    %h = urem i32 %a, %b
    %i = srem i32 %a, %b
    %j = shl i32 %a, %b
    %k = lshr i32 %a, %b
    %l = ashr i32 %a, %b
    %m = and i32 %a, %b
    %n = or i32 %a, %b
    %o = xor i32 %a, %b
    ret i32 0

  define i32 @mulEx(i32 %a, i32 %b) do
    %d = mul i32 %a, %b
    ret i32 0

  define float @fopsEx(float %a, float %b) do
    %c = fadd float %a, %b
    %d = fsub float %a, %b
    %e = fmul float %a, %b
    %f = fdiv float %a, %b
    %g = frem float %a, %b
    ret i32 0


#dump ops