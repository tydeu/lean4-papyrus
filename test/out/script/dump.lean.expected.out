; ModuleID = 'hello'
source_filename = "hello"

@0 = private unnamed_addr constant [12 x i8] c"Hello World\00", align 1

declare i8 @printf(i8* %0, ...)

define i32 @main() {
  %1 = call i8 (i8*, ...) @printf(i8* getelementptr inbounds ([12 x i8], [12 x i8]* @0, i32 0, i32 0))
  ret i32 0
}
void
i32
i8 (i8*, ...)
i8 addrspace(5)*
{ i32, float }
[2 x i8]
<4 x i64>
<vscale x 4 x double>
i1 true
i1 false
i8 -1
i32 1
i64 -1
i128 1208925819614629174706188
i128 -1208925819614629174706188
; ModuleID = 'ops'
source_filename = "ops"

define i32 @opsEx(i32 %a, i32 %b) {
  %c = add i32 %a, %b
  %d = mul i32 %a, %b
  %e = sub i32 %a, %b
  %f = udiv i32 %a, %b
  %g = sdiv i32 %a, %b
  %h = urem i32 %a, %b
  %i = srem i32 %a, %b
  %j = shl i32 %a, %b
  %k = lshr i32 %a, %b
  %l = ashr i32 %a, %b
  %m = and i32 %a, %b
  %n = or i32 %a, %b
  %o = xor i32 %a, %b
  ret i32 0
}

define i32 @mulEx(i32 %a, i32 %b) {
  %d = mul i32 %a, %b
  ret i32 0
}

define float @fopsEx(float %a, float %b) {
  %c = fadd float %a, %b
  %d = fsub float %a, %b
  %e = fmul float %a, %b
  %f = fdiv float %a, %b
  %g = frem float %a, %b
  ret i32 0
}
